src/mpid/ch4/netmod/ofi/ofi_win.h

```diff
@@ -949,8 +949,21 @@ static inline int MPIDI_NM_mpi_win_allocate_shared(MPI_Aint size,
     if (total_size == 0)
         goto fn_zero;
 
-    sprintf(shm_key, "/mpi-%X-%" PRIx64, MPIDI_Global.jobid, MPIDI_OFI_WIN(win).win_id);
+    /* Note that two distinct process groups may share the same context id,
+     * thus the window id may be duplicated on a node if windows are owned
+     * by distinct process groups. Therefore, we use [jobid + root_rank + win_id]
+     * as unique shm_key for window, where root_rank is the world rank of the
+     * first process in the group. */
+    int root_rank = 0;
+
+    if (comm_ptr->rank == 0)
+        root_rank = MPIR_Process.comm_world->rank;
+    mpi_errno = MPIR_Bcast_impl(&root_rank, 1, MPI_INT, 0, comm_ptr, &errflag);
+    if (mpi_errno != MPI_SUCCESS)
+        goto fn_fail;
```

Comment:
The `context_id` is a unique id for each communicator on the same process, as defined in MPI standard. So this means if I have two communicators, each includes totally distinct processes (e.g., comm0: [P0, P1], comm1: [P2, P3] ), they may have the same `context_id`.  But if the two communicators contain the same or some overlapped processes (e.g., comm0: [P0, P1], comm1: [P1, P2] ), they have to use different `context_id`, otherwise the semantics breaks on P1.

Now we add `root_rank` here, which is the world rank of one process in the communicator, as the second example shows, if two communicators have the same `root_rank`, they have to use different `context_id`. :)